File: khf1.erl
-module(khf1).
-author('csaba215@gmail.com').
-vsn('2020-10-02').
-export([cella/2]).
%-compile(export_all).

-type field() :: integer().
-type board() :: [[field()]].

-spec getcolumn(S::board(), RowStart::integer(), Length::integer(), Col::integer(), Out::[field()]) -> [field()].
getcolumn(_,1,0,_,Out) -> Out;
getcolumn([H|T],1,Length,Col,Out) -> getcolumn(T,1,Length-1,Col,[lists:nth(Col,H)|Out]);
getcolumn([_|T],RStart,Length,Col,Out) -> getcolumn(T,RStart-1,Length,Col,Out).

-spec getcolumns(S::board(), RowStart::integer(), RowLength::integer(), ColStart::integer(), ColLength::integer() , Out::[field()]) -> [field()].
getcolumns(_S, _RowStart, _RowLength, _ColStart, 0, Out) -> Out;
getcolumns(S, RowStart, RowLength, ColStart, ColLength, Out) -> Temp = getcolumn(S, RowStart, RowLength, ColStart, Out), getcolumns(S, RowStart, RowLength, ColStart+1, ColLength-1, Temp).

-spec khf1:cella(S::board(), I::integer()) -> C::[field()].
%% C egy olyan lista, amely az S Sudoku-mátrix I-edik
%% cellájának elemeit oszlopfolytonosan tartalmazza.
cella([H|T], I) -> K=trunc(math:sqrt(length(H))), lists:reverse(getcolumns([H|T], ((I-1) div K)*K+1, K, ((I-1) rem K)*K+1 ,K,[])).

File: khf1.pl
:- use_module(library(lists)).


%getcol([],Col,Start,Length,Out):-!.
getcol(_,_,_,0,[]):-!.
getcol([H|T],Col,1,Length,[R|Out]):- Length1 is Length-1,nth1(Col,H,R), getcol(T,Col,1,Length1,Out).
getcol([_|T],Col,Start,Length,Out):-
    Start1 is Start-1,
    getcol(T,Col,Start1,Length,Out).

%getcell([],ColStart,ColLength,RowStart,RowLength,Out):- !.
getcell(_,_,0,_RowStart,_RowLength,[]):-!.
getcell(M,Col,ColLength,RowStart,RowLength,Ret):- 
    getcol(M,Col,RowStart,RowLength,R),ColLength1 is ColLength-1,Col1 is Col+1,
    getcell(M,Col1,ColLength1,RowStart,RowLength,Out),append(R,Out,Ret).

% :- type field == int.
% :- type board  == list(list(field)).
% :- pred cella(board::in, int::in, list(field)::out).
% cella(S, I, C): C egy olyan lista, amely az S Sudoku-mátrix
%    I-edik cellájának elemeit oszlopfolytonosan tartalmazza.
cella(S, I, C):- 
    length(S,M),
    K is integer(sqrt(M)),
    RowStart is div(I-1,K)*K+1,
    ColStart is rem(I-1,K)*K+1,
    getcell(S,ColStart,K,RowStart,K,C).
